### ISSUE TITLE
listen() does not provide unique ports anymore

### ISSUE THREAD

**Issue by zaitcev** [eventlet/eventlet#411]
Tests in OpenStack gate started to fail sporadically with an extremely weird symptom: two services listening on the same port number. As we found, eventlet.listen(('localhost',0)) does not work like a normal bind to port zero would, e.g. it does not guarantee a unique port -- anymore. The following commit is the culprit:
https://github.com/eventlet/eventlet/commit/f9a3074a3b75f17f76cc04a693dc48a367b99861

Just in case, here's a test program:

```
#!/usr/bin/python

from __future__ import print_function
import sys
from eventlet.green import socket

def listen_zero():
    """
    The eventlet.listen() always sets SO_REUSEPORT, so when called with
    ("localhost",0), instead of returning unique ports it can return the
    same port twice.
    """
    # We cannot test against stock eventlet because it has no SO_REUSEPORT
    # in 0.18.4 on RHEL 7. Need to inline from trunk eventlet.
    # sock = eventlet.listen(("127.0.0.1", 0))
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
    sock.bind(("127.0.0.1", 0))
    sock.listen(50)
    return sock

def main():
    cnt = 0
    a = {}
    while cnt < 65000:
        sock = listen_zero()
        port = sock.getsockname()[1]
        if port in a:
            print("Port %d dup in %d tries\n" % (port, cnt))
            return 0
        # this also keeps port open
        a[port] = sock
        cnt += 1
    print("No dup in %d tries\n" % (cnt,), file=sys.stderr)
    return 1

if __name__ == '__main__':
    sys.exit(main())
```

I'm submitting an issue instead of a patch because I am not sure how to fix this. The best for me would be to back out the offending commit, but then Mr. Zhengwei Gao surely had a use case for it.

**Comment by zaitcev** [comment]
Here's the workaround we might use:
 https://review.openstack.org/463876

**Comment by temoto** [comment]
It's obviously a useful option (graceful restart with separate process binding to same socket), perhaps it should be disabled by default though.

**Comment by temoto** [comment]
> `    # We cannot test against stock eventlet because it has no SO_REUSEPORT`

No problem, you can, it's very easy and works on any system, even RedHat. I recommend ignore system packages and always install latest eventlet this or similar way.

```
virtualenv v
v/bin/pip install https://github.com/eventlet/eventlet/archive/master.zip
# or to test against particular version
v/bin/pip install -e git+https://github.com/eventlet/eventlet.git@8711100eab0af1bcd5cc5920741296c52d98d55b#egg=eventlet
# test
```

**Comment by temoto** [comment]
@zaitcev I really believe that this line in your `listen_zero()` function would help, it alleviates 2 minute (or so) wait after listening socket is closed, so address pair can be reused instantly.

**Comment by zaitcev** [comment]
A graceful restart has nothing to do with binding to port 0, when kernel allocates a unique port atomically. By definition it's going to be the new port every time. Graceful restart can only be a consideration when port is known ahead of time.

**Comment by temoto** [comment]
Thank you, I didn't think that way, that's absolutely correct, port 0 + REUSE_PORT is only a recipe for disaster.

**Comment by temoto** [comment]
Please, try again with this patch.
`SO_REUSEPORT` should still be on by default but only for non-zero port.
Both reuse options are now also controlled by kwargs and warning is issued for `reuse_port=True` with random port.

`pip install https://github.com/eventlet/eventlet/archive/0ec4df6cbe1771493d112c3f893b8ac1d89b5dbf.zip`

**Comment by clayg** [comment]
@temoto yes, this change works great - thanks!

**Comment by temoto** [comment]
It's merged into master 0ec4df6cbe1771493d112c3f893b8ac1d89b5dbf

@gaozhengwei please note `eventlet.listen()` now sets `SO_REUSEPORT` by default only when port is not 0 (random). Also it's possible to control reuseaddr/port options via kwargs.
